sys/net/link_layer/l2util/l2util.c

```diff
@@ -136,7 +136,6 @@ int l2util_eui64_from_addr(int dev_type, const uint8_t *addr, size_t addr_len,
             LOG_ERROR("l2util: can't convert hardware address to EUI-64 "
                       "for device type %d\n", dev_type);
 #endif  /* DEVELHELP */
-            assert(false);
```

Comment:
Huh? The program is supposed to abort here with `NDEBUG=0`.

Reply:
If this is supposed to be a utility function it should not abort the program when it is called with an invalid parameter, but return `-ENOTSUP`  - than the caller can `assert()` that the return value is as expected.

Reply:
You are correct that removing this assert embiggens the code quite a bit though :confused:

Reply:
`assert(false);` seems like a misuse of `assert`. This should probably `panic`. Since assert shall do nothing when `NDEBUG ` is defined.

Reply:
And why can't it be here? We already know that the device is not supported and just not failing usually leads to confused users why we do not get (in this case) an IPv6 address for the interface.

Reply:
But if it is your intention to change the call logic for `l2util`, this should rather be its own PR, as it seems to have nothing to do with this PR. Unless this removal is necessary, because this PR changes the semantics of `assert()`... then I question the validity of this PR.

Reply:
To quote @kfessel:

> Since assert shall do nothing when `NDEBUG ` is defined.

As far as I understand @benpicco's last comment, this changes that behavior to not return when `NDEBUG`. The point of assert is to be a debug and testing feature and it should not have any consequence in production code.

Reply:
This is definitely not correct. Either the API allows for calling with `dev_type` having an unsupported value, or not. If it is not allowed to call with an unsupported value, it is valid to have an `assert(false);` there, but the test application contains a bug that needs fixing: It calls the function with an forbidding value for `dev_type`.

If instead calling with an unsupported `dev_type` is allowed the test application is correct but the `assert(false);` needs to be dropped.

A bug-free program must not show different observable behavior depending on whether or not `NDEBUG`. The test app however does, so there definitely is a bug. The question is only if the caller or the called function needs to be fixed.

Reply:
assert with NDEBUG has an exact definition in the standard (`(void) 0`) sadly this definition leads to different compile results (for code that is not inside the assert) even with the most basic optimization enabled than without NDEBUG. @benpicco s change fixes that, i think it is legitimate to say most programmers want to debug the program that is running without debug being enabled.

`assert(false)` is a misuse of `assert` it is allways failing to program stop  ->  just call a `abort`/`panic` or `exit`
`assert(initilise_somthing)` is a misuse of `assert` assert should just contain tests ->` {int r = initilise_somthing; assert(r);}` this misuse will be hidden by this PR.
`assert(some arg is within the interface bounds)` is a legitimate use of  assert that currently triggers the unused warning  with NDEBUG.

In Total even though this does not follow the C standard to the letter, I think this patch is an improvement especially 
since we assume to always try to hint the optimizer to better code generation, if we do not do this PR because of the C-Standard we should consider adding a very similar `riot_assert` and replace most of our `asserts`.

Reply:
Ok, with your argumentation, I am fine with removal, but only if and only if an assert on `!= -ENOTSUP` is added for the result of any production code call of the removed asserts here is introduced.

Reply:
Sure it's only used in [3 places](https://github.com/RIOT-OS/RIOT/pull/19337/commits/0b0bc7173652ab30d5223cba46e6bf31be58a752) outside of tests.

Should the assert catch all errors or just `-ENOTSUP`?